src/opencl_common.c

```diff
@@ -2829,6 +2883,296 @@ int opencl_calc_min_kpc(size_t lws, size_t gws, int v_width)
 	return gws * v_width;
 }
 
+static void buf_list_init(struct opencl_buffers **list)
+{
+	*list = mem_alloc(sizeof(struct opencl_buffers));
+	(*list)->tail = (*list)->head = NULL;
+	(*list)->count = 0;
+}
+
+static void buf_list_add_link(struct opencl_buffers *list, struct buffer_opencl *entry)
+{
+	entry->next = NULL;
+
+	if (list->tail)
+		list->tail = list->tail->next = entry;
+	else
+		list->tail = list->head = entry;
+
+	list->count++;
+}
+
+static void buf_list_add(struct opencl_buffers *list, cl_mem *data, void **host_ptr)
+{
+	struct buffer_opencl *entry;
+
+	entry = mem_alloc(sizeof(struct buffer_opencl) + sizeof(data));
+	entry->data = *(data);
+	entry->variable = data;
```

Comment:
I think the ` + sizeof(data)` merely wastes memory - we never actually write beyond the size of the struct here, right?

Also, why do we store both `*data` and `data`? In other words, we store the same thing by value and by reference. Why?

Finally, assuming we need it, lets drop the braces in `*(data)`.

Reply:
> Also, why do we store both *data and data? In other words, we store the same thing by value and by reference. Why?

Think about 

1. the OpenCL buffer, the buffer itself, and;
2. the place where the format that made the call stores the information. 

What is of interest to the shared code is the OpenCL buffer, the reference is only for cleaning (assign NULL).